### PR TITLE
Allow thread_local!() to take attributes

### LINKED ISSUE #30756: thread_local!() can't take attributes
```
#![deny(unsafe_code)]

thread_local!(static KEY: u8 = 1);

fn main() {
}
```

will hit an unsafe_code warning due to an unsafe function internally defined in the thread local macro (the unsafe function is a recent addition). There's no way to `allow()` it.

The macro should take in attributes.

### PR DESCRIPTION
**Pull request by Manishearth**
fixes #30756

r? @Gankro

### CODE REVIEW

**Comment by Gankro** [comment]
r? @brson 

(I have no background on TLS/macros/attributes)

**Comment by alexcrichton** [comment]
As mentioned in #30756 my preferred "fix" for now would be to not change the public interface and instead add some `#[allow]` attributes for the lints that may get triggered. I'd want to think a little more before changing the public interface here personally

**Comment by Manishearth** [comment]
Updated so that it allows unsafe/dead code

**Comment by Manishearth** [comment]
er, git push stopped working, updated now

**Comment by alexcrichton** [comment]
Can you add a test for this as well?

**Comment by Manishearth** [comment]
Done

File: src/test/run-pass/issue-30756.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![deny(unsafe_code)]
+
+thread_local!(static FOO: u8 = 1);
+
+fn main() {
+}
```
Comment (eddyb) [inline src/test/run-pass/issue-30756.rs]: Missing newline.

**Comment by brson** [comment]
Doesn't putting `allow` attributes in this macro make it unusable for anybody applying `forbid(unused_code)` or `forbid(dead_code)`?

**Comment by brson** [comment]
r-

Per my previous comment, putting `allow` attributes in this macro makes it unusable with `forbid`. This undermines the usefulness of forbid. Are there other places where we are embedding lint directives in macros?

**Comment by brson** [comment]
@bors r-

Per my previous comment, putting `allow` attributes in this macro makes it unusable with `forbid`. This undermines the usefulness of forbid. Are there other places where we are embedding lint directives in macros?

**Comment by Manishearth** [comment]
Yeah, it does. I'd fixed it in a more flexible way in d02f969 where attributes can be passed in, if that is more amenable.

**Comment by alexcrichton** [comment]
Hm I thought that we had other examples of macros that did things like this, but we apparently don't. In retrospect I believe `allow(dead_code)` should be removed because all names are prefixed with underscores, which should prevent that lint from triggering anyway.

**Comment by Manishearth** [comment]
A bunch of macros like bitflags do the thing where they can accept attributes. I don't see any which contain allows.

**Comment by alexcrichton** [comment]
The libs team discussed this during triage today and the conclusion was that the easiest fix here was just removing the unsafe function entirely. @Manishearth would you be ok updating the patch to just changing it to a safe function? The unsafety here is all synthetic as it's all an implementation detail anyway.

**Comment by Manishearth** [comment]
Done. I'm not sure where the unsafe boundary should eventually end up, so you may want to have a close look.

**Comment by alexcrichton** [comment]
Ah well technically this *was* the boundary (a function returning a `'static` reference which wasn't actually `'static`), but oh well.

Looks like the travis failure is legit (needs an unsafe block around a call to an intrinsic), but otherwise lgtm

**Comment by Manishearth** [comment]
Yeah, I fixed the travis failure locally, but tests haven't run yet.

**Comment by Manishearth** [comment]
Tests pass. r?

**Pull request merged**